pkg/identity/cidr.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package identity
+
+import (
+	"net"
+
+	"github.com/cilium/cilium/pkg/labels/cidr"
+)
+
+// AllocateCIDRIdentities allocates identities for each of the specified CIDR
+// prefixes so they will be available later during policy resolution.
+//
+// On success, the returned slice will have a 1-to-1 correspondence with the
+// CIDR in the provided prefixes slice, eg the identity for prefixes[0] will
+// be held in identities[0].
+// On failure, attempts to clean up after itself and returns the error.
+func AllocateCIDRIdentities(prefixes []*net.IPNet) (res []*Identity, err error) {
+	if len(prefixes) == 0 {
+		return nil, nil
+	}
+
+	log.Debugf("Attempting to allocate identities for %d prefixes", len(prefixes))
+
+	// Allocate identities for each CIDR.
+	res = make([]*Identity, 0, len(prefixes))
+	for _, prefix := range prefixes {
+		if prefix == nil {
+			continue
+		}
+		labels := cidr.GetCIDRLabels(prefix)
+		id, _, err := AllocateIdentity(labels)
+		if err != nil {
+			// If any identity allocation failed, release existing identities
+			// and log the error.
+			if err2 := ReleaseSlice(res); err2 != nil {
+				log.WithError(err2).Error("Could not recover from error during CIDR identity allocation")
+			}
+			res = nil
+			break
+		}
+		res = append(res, id)
+	}
+
+	if err == nil {
+		log.Debugf("Allocated identities for %d prefixes", len(res))
+	} else {
+		log.WithError(err).Warningf("Failed to allocate identities for %d prefixes", len(res))
+	}
+
+	return res, err
+}
+
+// LookupCIDRIdentities finds identities corresponding to each of the specified
+// prefixes. It expects to be able to find all identities.
+// On success, returns a slice of identities with a 1-to-1 correspondence to
+// the specified slice of prefixes, and nil.
+// On error, returns all identities that can be resolved and an error.
+func LookupCIDRIdentities(prefixes []*net.IPNet) (res []*Identity, err error) {
+	res = make([]*Identity, 0, len(prefixes))
+	err = nil
```

Comment:
since this is a named return variable, it's already nil (the 0 value for an `error`)

Reply:
While debugging at one point, I was beginning to doubt this..